### PR TITLE
Update max_version on yank

### LINKED ISSUE #76: Yanking a crate should update its max_version
If a crate has its max version yanked, it should update the max_version field to go back to the otherwise maximum version.

### PR DESCRIPTION
**Pull request by yodaldevoid**
Fixes #76

### CODE REVIEW

**Comment by yodaldevoid** [comment]
One thing to note, these changes have yet to be tested with the front end. The only thing that comes to mind as a possible problem is the fact that max_version can now be expected to be null. It was never restricted to not being null, but it was never a true possibility due to the backend code.

**Comment by yodaldevoid** [comment]
I have looked into testing the frontend response to a null max_value for a crate. The only test I can think of is to start up a local backend and frontend, publish a crate to the local backend, yank the crate from the local backend, and finally try to view it on the frontend. My problem comes from trying to publish to the local backend using `cargo`. I have added the config block spat out by script/init-local-index.sh, but cargo always fails with a 404 error if `--host http://127.0.0.1:8888/` is passed to `cargo` and an SSL error if `--host https://127.0.0.1:8888/` is passed to `cargo`. Is there something I could be forgetting, or am I simply going about this wrong?

**Comment by carols10cents** [comment]
> I have added the config block spat out by script/init-local-index.sh, but cargo always fails with a 404 error if --host http://127.0.0.1:8888/ is passed to cargo and an SSL error if --host https://127.0.0.1:8888/ is passed to cargo. Is there something I could be forgetting, or am I simply going about this wrong?

Ugh yeah, `--host` isn't actually host of the crates server, it wants the location of the index :( When running locally with crates.io setup as recommended in the readme, this value should be something like `file:///path/to/your/crates.io/checkout/tmp/index-co/`. This has bothered me for a while, I'm going to go file a bug with cargo right now.

The next problem that you'll probably run into is that, unless you've set up an s3 bucket, publishing to your local crates.io will fail trying to upload the code to s3. I get around this by temporarily applying [this commit](https://github.com/integer32llc/crates.io/commit/4633bb5d67017a1673eeb8518dd177be41ca9d68) to disable uploads to s3 and just update the database-- this has been bothering me as well and I want to fix it as part of #470.

Basically I'm sorry i'm sorry.

**Comment by carols10cents** [comment]
Alternately, I'm happy to test this out on my setup, since I've already gone through this pain. Let me know!! I'm looking at the code now :)

**Comment by carols10cents** [comment]
The code is looking great!! 

Other than checking out the frontend, the only other thing this change will need is a way to go through existing crates and update those whose max_versions have been yanked to have their next highest non-yanked version used (or set to null if all versions have been yanked) as [retep998 reminded me](https://github.com/rust-lang/crates.io/issues/76#issuecomment-281502943) 

I should have elaborated in the issue, but we have some one-off tasks in src/bin, such as [this task that backfilled users' github IDs](https://github.com/rust-lang/crates.io/blob/master/src/bin/fill-in-user-id.rs) (ignore the docs at the top of that file, they're copy-pasted from another binary).

If you don't want to take care of that for whatever reason, I'm happy to, let me know!

**Comment by yodaldevoid** [comment]
Thanks for pointing me in the right direction! Don't worry about the problems, now we know more about what needs to be improved. I'll be honest, whenever I find a place where documentation is lacking when working with Rust I feel a little better as it reminds me that everyone working on this language/project is human.

It's interesting that `--host` is used for publishing, but `--index` (which honestly makes more sense) is used for yanking. At any rate, here is the front page of a crate with a null max_version.
null_crate
I don't think the red bar at the top is acceptable, but everything else seems fine to me.

I'll look at writing the one off task. It should be fairly easy to make as most of the logic is already in `yank`.

**Comment by yodaldevoid** [comment]
Hmmm, this is a little concerning. I tested publishing a newer version of a crate when the max version was null, and it seems that the max version is not being updated. I'll look at adding a test and code to fix this.

**Comment by carols10cents** [comment]
Everything you've said sounds great! Also I [filed a bug with cargo about the `--host` flag.](https://github.com/rust-lang/cargo/issues/3797)

> I'll be honest, whenever I find a place where documentation is lacking when working with Rust I feel a little better as it reminds me that everyone working on this language/project is human.

We are all very much human :) ❤️

**Comment by yodaldevoid** [comment]
Everything should be wrapped up. I tested it against my local copy and it worked. I don't have tests for the update task, but can add those if you wish.

**Comment by yodaldevoid** [comment]
Going through the issues I found a couple more issues that this addresses. #510 might be fixed, but there might be some more work to be done there.

Fixes #502

**Comment by sgrif** [comment]
Just to note, #592 is an alternate solution to this.

**Comment by carols10cents** [comment]
Oh geez. Before merging either this or #592, I want to hear from alex why max_version was added in the first place. And yes, the tests in here are valuable no matter what ❤️

**Comment by yodaldevoid** [comment]
I have added tests to the one-off.

**Comment by carols10cents** [comment]
@yodaldevoid thank you so much for your work on this, and I hate to do this-- I really should have questioned whether we actually need this column like @sgrif did in #592. I'm definitely going to merge your tests in, which will be valuable in making sure the code in #592 works, but if it does and if production performs ok with the max_version column removed, we might not need this code. 

But I want to reassure you that your code is great and I'm very grateful!! ❤️

**Comment by yodaldevoid** [comment]
No problem, really! I'm all for simplifying the database.

**Comment by carols10cents** [comment]
Thank you for understanding!!!

**Comment by yodaldevoid** [comment]
Closing because #592
